kv,kvcoord,sql: poison txnCoordSender after a retryable error

lidorcarmel opened the pull request:
Previously kv users could lose parts of a transaction without getting an
error. After Send() returned a retryable error the state of txn got reset
which made it usable again. If the caller ignored the error they could
continue applying more operations without realizing the first part of the
transaction was discarded. See more details in the issue (#22615).

The simple case example is where the retryable closure of DB.Txn() returns
nil instead of returning the retryable error back to the retry loop - in this
case the retry loop declares success without realizing we lost the first part
of the transaction (all the operations before the retryable error).

This PR leaves the txn in a "poisoned" state after encountering an error, so
that all future operations fail fast. The caller is therefore expected to
reset the txn handle back to a usable state intentionally, by calling
Txn.PrepareForRetry(). In the simple case of DB.Txn() the retry loop will
reset the handle and run the retry even if the callback returned nil.

Closes #22615

Release note: None

Commit:
kv,kvcoord,sql: poison txnCoordSender after a retryable error

Previously kv users could lose parts of a transaction without getting an
error. After Send() returned a retryable error the state of txn got reset
which made it usable again. If the caller ignored the error they could
continue applying more operations without realizing the first part of the
transaction was discarded. See more details in the issue (#22615).

The simple case example is where the retryable closure of DB.Txn() returns
nil instead of returning the retryable error back to the retry loop - in this
case the retry loop declares success without realizing we lost the first part
of the transaction (all the operations before the retryable error).

This PR leaves the txn in a "poisoned" state after encountering an error, so
that all future operations fail fast. The caller is therefore expected to
reset the txn handle back to a usable state intentionally, by calling
Txn.PrepareForRetry(). In the simple case of DB.Txn() the retry loop will
reset the handle and run the retry even if the callback returned nil.

Closes #22615

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74563)
<!-- Reviewable:end -->

nvb reviewed (commented):
Nice job! It will be a big step forward to finally resolve this issue.

> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-:-MtlSitl0jkLzUMWhpp2:bqzhtcb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @lidorcarmel)

___
*[pkg/kv/txn.go, line 978 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlM8nw0VA07SsLO0Ry:-MtlM8nx0Grn2d8LHMry:b82mygr) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/txn.go#L978)):*
> ```Go
> 		if err == nil {
> 			txn.mu.Lock()
> 			txnErr := txn.mu.sender.GetTxnRetryableErr(ctx)
> ```

Why do we need to explicitly check this here? Shouldn't the call to `txn.Commit()` below and the logic in `maybeRejectClientLocked` be enough to detect a poisoned txn and return an error?
___
*[pkg/kv/txn.go, line 1421 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlNBVZ3nw4sRYKp3Ei:-MtlNBVZ3nw4sRYKp3Ej:b-hca1pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/txn.go#L1421)):*
> ```Go
> 	}
> 	// Verify we have an initialized Transaction before creating a new root sender.
> 	if newTxn.ID == (uuid.UUID{}) {
> ```

Is this a valid case to hit?
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlNPCs5qjxSuDG_YTx:-MtlNPCs5qjxSuDG_YTy:b-2yyho2) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L115)):*
> ```Go
> 		txnState txnState
> 
> 		// storedRetryableErr is set when tsnState == txnRetryableError. This
> ```

`s/tsnState/txnState/`
___
*[pkg/sql/conn_fsm.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlPA8P-IKtYynNU79e:-MtlPA8P-IKtYynNU79f:bdxj8om) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/sql/conn_fsm.go#L377)):*
> ```Go
> 			Description: "ROLLBACK TO SAVEPOINT (not cockroach_restart) failed because txn needs restart",
> 			Next:        stateAborted{},
> 			Action:      prepareTxnForRetry,
> ```

Are you sure this is where we wanted the call to `PrepareForRetry`, as opposed to the next event which handles the `cockroach_restart` savepoint and advances the `txnState` with a `txnRestart` event? I'm a little surprised to not see a 1:1 mapping between uses of `txnRestart` and calls to `PrepareForRetry`.

Is this related to the need to provide an error to `PrepareForRetry`?
___
*[pkg/kv/db_test.go, line 724 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlQYc13FrevcTOXvDN:-MtlQYc13FrevcTOXvDO:bdahg8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/db_test.go#L724)):*
> ```Go
> 			// Only write if we have not written before, because otherwise we will keep aborting
> 			// the other txn forever.
> 			r, e := hpTxn.Get(ctx, "aa")
> ```

nit: `s/e/err/` throughout.
___
*[pkg/kv/db_test.go, line 755 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlR16p9PtF2a-ICppy:-MtlR16p9PtF2a-ICppz:bkbumfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/db_test.go#L755)):*
> ```Go
> 		return e
> 	})
> 	require.NoError(t, err)
> ```

Want to add `require.Equal(t, 1, runNumber)` here to assert that we performed exactly one retry?
___
*[pkg/kv/db_test.go, line 758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlQpyiDwRNfEl5CufV:-MtlQpyiDwRNfEl5CufW:bsunm63) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/db_test.go#L758)):*
> ```Go
> 
> 	err1 := db.Txn(context.Background(), func(ctx context.Context, txn *kv.Txn) error {
> 		// The high priority txn was overridden by the successful  retry.
> ```

nit: extra space in "successful retry". Also, I think `s/overridden/overwritten/` is slightly more accurate.
___
*[pkg/kv/db_test.go, line 822 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlRPw-EThnBLN-_SIM:-MtlRPw-EThnBLN-_SIN:b-vocyw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/db_test.go#L822)):*
> ```Go
> 	// db.Txn should return the retryable error that poisoned txn.
> 	expectedErr := (*roachpb.TransactionRetryWithProtoRefreshError)(nil)
> 	require.True(t, errors.As(err, &expectedErr))
> ```

Same thing here. Consider a `require.Equal(t, 0, runNumber)`.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 705 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlRrvK6q9FBQf-Hm7L:-MtlRrvK6q9FBQf-Hm7M:b-4jnoqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L705)):*
> ```Go
> 		{
> 			// On abort, nothing changes but we get a new priority to use for
> 			// the next attempt.
> ```

It looks like we broke what the test was trying to demonstrate. Can we revive it by adding a call to `PrepareForRetry`?


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
thanks and also thanks for the chat about this PR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-:-MtugKgZ3PNRDxOtsNgV:b-n5sjc7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/kv/txn.go, line 978 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlM8nw0VA07SsLO0Ry:-Mtpoj1j9YeSV-xRfitv:bj6g9yv) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/txn.go#L978)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we need to explicitly check this here? Shouldn't the call to `txn.Commit()` below and the logic in `maybeRejectClientLocked` be enough to detect a poisoned txn and return an error?
</blockquote></details>

as we talked, I dropped these lines and documented that in DB.Txn().
note that now we only need one unit test instead of 2 - because the behavior is the same whether the callback returns nil or the error. the new test is less cool but I'll live with it :)
thanks!
___
*[pkg/kv/txn.go, line 1421 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlNBVZ3nw4sRYKp3Ei:-MtuctJ1BEV85IHlXDxv:bjc0y47) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/txn.go#L1421)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this a valid case to hit?
</blockquote></details>

it was but as we discussed, I removed those cases and dropped this.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlNPCs5qjxSuDG_YTx:-Mtpp8D73JcbY-U82mxK:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L115)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/tsnState/txnState/`
</blockquote></details>

done.
___
*[pkg/sql/conn_fsm.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlPA8P-IKtYynNU79e:-Mtt45cO7J_i-k6IH4Bu:b3qvhn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/sql/conn_fsm.go#L377)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are you sure this is where we wanted the call to `PrepareForRetry`, as opposed to the next event which handles the `cockroach_restart` savepoint and advances the `txnState` with a `txnRestart` event? I'm a little surprised to not see a 1:1 mapping between uses of `txnRestart` and calls to `PrepareForRetry`.

Is this related to the need to provide an error to `PrepareForRetry`?
</blockquote></details>

interesting. I was thinking that we are doing `PrepareForRetry` where we have `eventRetriableErr` and we still expect to use the handle (otherwise we can just do `CleanupAndFinishOnErr`).
WDYT?
___
*[pkg/kv/db_test.go, line 724 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlQYc13FrevcTOXvDN:-MtpqicRCKXb_ziegicK:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/db_test.go#L724)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `s/e/err/` throughout.
</blockquote></details>

done.
___
*[pkg/kv/db_test.go, line 755 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlR16p9PtF2a-ICppy:-MtpqWupA1zUk90cZIHa:b-fm6nat) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/db_test.go#L755)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Want to add `require.Equal(t, 1, runNumber)` here to assert that we performed exactly one retry?
</blockquote></details>

yes, done (Equal 2).
___
*[pkg/kv/db_test.go, line 758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlQpyiDwRNfEl5CufV:-Mtppmu_CgUw51iQ9xtW:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/db_test.go#L758)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: extra space in "successful retry". Also, I think `s/overridden/overwritten/` is slightly more accurate.
</blockquote></details>

done.
___
*[pkg/kv/db_test.go, line 822 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlRPw-EThnBLN-_SIM:-MtpqUQH8fKbGFHL-QYu:b-g1dp6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/db_test.go#L822)):*
> require.Equal(t, 1, runNumber) 
deleted this test.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 705 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlRrvK6q9FBQf-Hm7L:-Mtsz2Cu2EbcJ5Y3BSXn:b4ul50y) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L705)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It looks like we broke what the test was trying to demonstrate. Can we revive it by adding a call to `PrepareForRetry`?
</blockquote></details>

it was intentional but I definitely missed the comment.
how about this? having one test case for aborting and one for aborting and reseting the handle?



_Code quote:_
```Go
		// Is set, we're expecting that the Transaction proto is re-initialized (as
		// opposed to just having the epoch incremented).
		expNewTransaction bool
```


<!-- Sent from Reviewable.io -->

lidorcarmel commented:
Hi @yuzefovich, hopefully you can review the SQL side of this PR.

nvb reviewed (commented):
> Reviewed 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @lidorcarmel, and @yuzefovich)

___
*[pkg/kv/db.go, line 825 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MuBgokcBaDw68ABiZ_z:-MuBgokcBaDw68ABiZa-:b-euhoy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e21d07088332cafd2009a4b0a1fa423c56c1155/pkg/kv/db.go#L825)):*
> ```Go
> //      return err
> //    }
> //    // ...
> ```

nit: the indentation of this looks off. Are we mixing spaces and tabs?
___
*[pkg/sql/conn_fsm.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlPA8P-IKtYynNU79e:-MuBhd9o0-pOzeFTULe3:b-2pb9a7) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/sql/conn_fsm.go#L377)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

interesting. I was thinking that we are doing `PrepareForRetry` where we have `eventRetriableErr` and we still expect to use the handle (otherwise we can just do `CleanupAndFinishOnErr`).
WDYT?
</blockquote></details>

Are these cases where we're calling `setAdvanceInfo` with `noEvent` and staying in `stateAborted` real retryable errors? The fact that this isn't moving us back to `stateOpen` is suspect to me. Where is this error being generated? Is this the psuedo-retryable error returned by `TxnCoordSender.RollbackToSavepoint`?
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 724 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MuBhM3qC11tIjT7pz-y:-MuBhM3qC11tIjT7pz-z:b-mjnhue) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e21d07088332cafd2009a4b0a1fa423c56c1155/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L724)):*
> ```Go
> 			// On abort, reset the txn by calling PrepareForRetry, and then we get a
> 			// new priority to use for the next attempt.
> 			name: "TransactionAbortedErrorWithPrepareForRetry",
> ```

nit: `s/TransactionAbortedErrorWithPrepareForRetry/TransactionAbortedError with PrepareForRetry/`


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Hm, I took at the changes in `sql` package, and unfortunately I'm not that familiar with the code being modified. Furthermore, probably nobody on the SQL teams is familiar with it either - it is kinda a no man's land because it was written by @andreimatei long time ago and we didn't have to make many changes to it. I'm guessing that Andrei doesn't have much context on it either :/

lidorcarmel reviewed (commented):
thank Yahor! I think Andrei has some context, I'll check with him.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-:-MuEUUVS8Ix3_0Ak9ynZ:b-q53ioc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @lidorcarmel, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/db.go, line 825 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MuBgokcBaDw68ABiZ_z:-MuDXoyB4E5_LzEfx6i2:b-dtkdrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e21d07088332cafd2009a4b0a1fa423c56c1155/pkg/kv/db.go#L825)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: the indentation of this looks off. Are we mixing spaces and tabs?
</blockquote></details>

Done, thanks.
___
*[pkg/sql/conn_fsm.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlPA8P-IKtYynNU79e:-MuEUxYr0Ar5fWEFx-Aa:b8mbh7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/sql/conn_fsm.go#L377)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are these cases where we're calling `setAdvanceInfo` with `noEvent` and staying in `stateAborted` real retryable errors? The fact that this isn't moving us back to `stateOpen` is suspect to me. Where is this error being generated? Is this the psuedo-retryable error returned by `TxnCoordSender.RollbackToSavepoint`?
</blockquote></details>

working.. leaving open.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 724 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MuBhM3qC11tIjT7pz-y:-MuDYlBs1Si6kA_szZh0:bsgzn6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e21d07088332cafd2009a4b0a1fa423c56c1155/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L724)):*
> TransactionAbortedError with PrepareForRetry
Done.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @lidorcarmel, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/conn_fsm.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlPA8P-IKtYynNU79e:-MuFpfrq0S13mjWs_pyt:b8a4fis) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/sql/conn_fsm.go#L377)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

working.. leaving open.
</blockquote></details>

Can you show us the generated diagram for the fsm after the change?


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @lidorcarmel, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/conn_fsm.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlPA8P-IKtYynNU79e:-MuJeO5NE2Uvb_RoSQKQ:brl9rid) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/sql/conn_fsm.go#L377)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you show us the generated diagram for the fsm after the change?
</blockquote></details>

Looks like I did not change the diagram because this PR only adds actions, so I guess we're still with https://github.com/cockroachdb/cockroach/blob/b1381aa9a779051c0a89af1d01142a150cefc0b9/pkg/sql/txnstatetransitions_diagram.png



_Code quote:_
```Go
				args.Extended.(*txnState).setAdvanceInfo(skipBatch, noRewind, noEvent)
				return nil
			},
```


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @lidorcarmel, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/conn_fsm.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlPA8P-IKtYynNU79e:-Mum9nciC_8_561fo2et:b-s6qvae) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/sql/conn_fsm.go#L377)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

Looks like I did not change the diagram because this PR only adds actions, so I guess we're still with https://github.com/cockroachdb/cockroach/blob/b1381aa9a779051c0a89af1d01142a150cefc0b9/pkg/sql/txnstatetransitions_diagram.png
</blockquote></details>

PTAL:
* changed conn_fsm to be less eager to PrepareForRetry as you suggested Nathan. Instead, preparing for retry when doing the restart and not when seeing the retriable error.
* PrepareForRetry doesn't require the caller to provide the error because the coordinator has it (this change is needed because conn_fsm doesn't know the error when doing the restart, and also this makes more sense this way..).


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-:-Muq_OaeCl-3MBpo6YOZ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @lidorcarmel, and @yuzefovich)

___
*[pkg/sql/conn_fsm.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlPA8P-IKtYynNU79e:-MuqZo41ERbvE3e4Gkft:b3s6ma) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/sql/conn_fsm.go#L377)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

PTAL:
* changed conn_fsm to be less eager to PrepareForRetry as you suggested Nathan. Instead, preparing for retry when doing the restart and not when seeing the retriable error.
* PrepareForRetry doesn't require the caller to provide the error because the coordinator has it (this change is needed because conn_fsm doesn't know the error when doing the restart, and also this makes more sense this way..).
</blockquote></details>

These changes look very good. Thank you for making them.


<!-- Sent from Reviewable.io -->

lidorcarmel commented:
TFTR Nathan!
bors r+

lidorcarmel commented:
retrying..
bors r+

lidorcarmel commented:
actually it might be related to this pr.
bors r-

lidorcarmel reviewed (commented):
hi @nvanbenschoten , looks like that test failure in rttanalysis was real, the issue was that when things are slow (for some reason only on bazel) then the txn may get aborted by a heartbeat failure, BUT, when we try to PrepareForRetry without an error we fail because we cannot find the error - a retryable error created by the heartbeat failure did not poison the handle, so it's a real bug! thanks bazel :)
I tried adding a test, please feel free to tell me that there is a much better/easier way to do this, but this test does the job.
BTW I think that regardless of this change we should call handleRetryableErrLocked() when we see that heartbeat failure.
Last - the lock/unlock stuff is not related to the test failure.
PTAL, thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-:-MvMrJ3W4ZkCgJ3Fg07K:bwnapbv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @lidorcarmel, @nvanbenschoten, and @yuzefovich)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 2 of 5 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @lidorcarmel, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 731 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOga1KCcqzlUZaOrGQ:-MvOga1KCcqzlUZaOrGR:b-fnf516) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L731)):*
> ```Go
> 		abortedErr := roachpb.NewErrorWithTxn(
> 			roachpb.NewTransactionAbortedError(roachpb.ABORT_REASON_CLIENT_REJECT), &tc.mu.txn)
> 		if tc.typ == kv.LeafTxn {
> ```

LeafTxns don't use `txnHeartbeaters`, so I don't see how this case could be possible. I think I just missed that when this was last touched in https://github.com/cockroachdb/cockroach/commit/1a84480056786eb3fc11e6bb4f35d3c229fbc53f. Do you mind instrumenting it with a fatal and seeing whether we ever hit it in unit tests? If not, I think we should remove it.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 737 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOgPaoE2kZvwcaCTJT:-MvOgPapD0j7peFc3VGN:b-2hhuku) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L737)):*
> ```Go
> 		}
> 		// Root txns handle retriable errors.
> 		return roachpb.NewError(tc.handleRetryableErrLocked(ctx, abortedErr))
> ```

This is a good catch. I agree with you that we should make this change. Can we do so in a separate commit though? That way, we can better pinpoint any fallout from it.
___
*[pkg/sql/conn_fsm.go, line 506 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOi6ajCzzhGqpZshK6:-MvOi6ajCzzhGqpZshK7:bfgbv3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/sql/conn_fsm.go#L506)):*
> ```Go
> 	ts := args.Extended.(*txnState)
> 	ts.mu.Lock()
> 	ts.mu.txn.PrepareForRetry(ts.Ctx)
> ```

I don't think it needs to hold up this PR, but one thing in here that I'd be interested to get @andreimatei's take on is whether it is more appropriate for these calls to `PrepareForRetry` to live in these fsm Action functions or whether they should be moved into `txnStateTransitionsApplyWrapper` and handled on the `txnRestart` event. I don't think I ever understood what belongs where and why all connExecutor state updates weren't handled directly in an fsm Action.
___
*[pkg/kv/db_test.go, line 17 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOjGs_EtWr0V_f58EH:-MvOjGs_EtWr0V_f58EI:bj8njow) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L17)):*
> ```Go
> 	"context"
> 	"fmt"
> 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
> ```

nit: move this into the grouping below. crlfmt should do this automatically, so I wonder whether you have that set up: https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/154206209/Goland+Tips+and+Tricks#Enable-crlfmt-Watcher. If not, I'd recommend it.
___
*[pkg/kv/db_test.go, line 802 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOl_AP1PB8A6gpekdL:-MvOl_AP1PB8A6gpekdM:bmn8a2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L802)):*
> ```Go
> }
> 
> // Verify the txn sees a retryable error without using the handle: Normally the
> ```

This might be a more appropriate test for `pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go`.
___
*[pkg/kv/db_test.go, line 810 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOknqlDoq-_37WWB8H:-MvOknqlDoq-_37WWB8I:b-35jetn) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L810)):*
> ```Go
> // handle stays in txnPending then this means we do not have the error. This
> // error is needed for resetting the handle in PrepareForRetry.
> func TestDB_PrepareForRetryAfterHeartbeatFailure(t *testing.T) {
> ```

How long does this test to run? At least 1 second, right? If you replace `setup` with a call to `serverutils.StartServer` directly, you should be able to configure a shorter txn heartbeat interval?
___
*[pkg/kv/db_test.go, line 833 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOl8Hc8n7IJuVwsdkl:-MvOl8Hc8n7IJuVwsdkm:ba58dct) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L833)):*
> ```Go
> 		// This is here because we want to call maybeRejectClientLocked, which will
> 		// poison the handle if there is a heartbeat failure.
> 		tc.GetLeafTxnFinalState(ctx, kv.OnlyPending)
> ```

nit: `_, _ = tc.GetLeafTxnFinalState(ctx, kv.OnlyPending)` to show that you're intentionally ignoring the result. CI might actually be complaining about this already.
___
*[pkg/kv/db_test.go, line 841 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOlhdO9bHvxcCqgr1p:-MvOlhdO9bHvxcCqgr1q:bcru90d) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L841)):*
> ```Go
> 			return errors.New("the handle is not poisoned yet")
> 		}
> 		return nil
> ```

Should we assert on the contents of the error after we've checked that it's non-nil?


<!-- Sent from Reviewable.io -->

Commit:
kvcoord: unify the way we handle retryable errors

Instead of calling handleRetryableErrLocked() we did similar work
when the transaction sees a retryable error from a heartbeat failure.

Instead, it's better to call the existing handleRetryableErrLocked()
to take care of those errors in one place.

Release note: None

Commit:
kvcoord: removing unnecessary handling in LeafTxn

LeafTxns don't use heartbeats, therefore we should never hit
this condition.

Release note: None

lidorcarmel reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @lidorcarmel, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 731 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOga1KCcqzlUZaOrGQ:-MvV7C-74gigNWF0Iny4:b-ahvjgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L731)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

LeafTxns don't use `txnHeartbeaters`, so I don't see how this case could be possible. I think I just missed that when this was last touched in https://github.com/cockroachdb/cockroach/commit/1a84480056786eb3fc11e6bb4f35d3c229fbc53f. Do you mind instrumenting it with a fatal and seeing whether we ever hit it in unit tests? If not, I think we should remove it.
</blockquote></details>

Tested locally with some unit tests, and also pushed to test in CI.
Removing this in a separate commit (I'm getting used to thinking about commits!).
CI is still running, if it fails I'll revert.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 737 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOgPaoE2kZvwcaCTJT:-MvV7MlOEzzmfXYUC-7A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L737)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is a good catch. I agree with you that we should make this change. Can we do so in a separate commit though? That way, we can better pinpoint any fallout from it.
</blockquote></details>

Done.
___
*[pkg/sql/conn_fsm.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MtlPA8P-IKtYynNU79e:-MvX9WksA6HJAzmQd83F:b-9755fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/39718c4a8b4c43f5c99c6ddddd067393e8706be0/pkg/sql/conn_fsm.go#L377)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

These changes look very good. Thank you for making them.
</blockquote></details>

Thank you!
___
*[pkg/sql/conn_fsm.go, line 506 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOi6ajCzzhGqpZshK6:-MvUxrR-4yWNio3Xcgq2:b-m24ezr) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/sql/conn_fsm.go#L506)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't think it needs to hold up this PR, but one thing in here that I'd be interested to get @andreimatei's take on is whether it is more appropriate for these calls to `PrepareForRetry` to live in these fsm Action functions or whether they should be moved into `txnStateTransitionsApplyWrapper` and handled on the `txnRestart` event. I don't think I ever understood what belongs where and why all connExecutor state updates weren't handled directly in an fsm Action.
</blockquote></details>

Looks like we're good here, verified with Andrei, conn_fsm should manage the txn object.
___
*[pkg/kv/db_test.go, line 17 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOjGs_EtWr0V_f58EH:-MvV7kQtAgdx6beDp66a:b-jukdp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L17)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: move this into the grouping below. crlfmt should do this automatically, so I wonder whether you have that set up: https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/154206209/Goland+Tips+and+Tricks#Enable-crlfmt-Watcher. If not, I'd recommend it.
</blockquote></details>

Done. Strange! I had crlfmt.. and I'm sure it was working until recently but now it was disabled.... mmm.
___
*[pkg/kv/db_test.go, line 802 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOl_AP1PB8A6gpekdL:-MvX0PKP3jjWUKkkp9wd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L802)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This might be a more appropriate test for `pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go`.
</blockquote></details>

Done.
___
*[pkg/kv/db_test.go, line 810 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOknqlDoq-_37WWB8H:-MvX0BGs2ApYL6628OId:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L810)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

How long does this test to run? At least 1 second, right? If you replace `setup` with a call to `serverutils.StartServer` directly, you should be able to configure a shorter txn heartbeat interval?
</blockquote></details>

Done.
___
*[pkg/kv/db_test.go, line 833 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOl8Hc8n7IJuVwsdkl:-MvWzUsXE1SKjwjLjifV:b-3jb9hr) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L833)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `_, _ = tc.GetLeafTxnFinalState(ctx, kv.OnlyPending)` to show that you're intentionally ignoring the result. CI might actually be complaining about this already.
</blockquote></details>

Done.
No idea why it didn't complain :(



_Code quote (from pkg/kv/kvclient/kvcoord/txn_coord_sender.go):_
```Go
		newTxn := roachpb.PrepareTransactionForRetry(
			ctx, abortedErr, roachpb.NormalUserPriority, tc.clock)
		return roachpb.NewError(roachpb.NewTransactionRetryWithProtoRefreshError(
			abortedErr.String(), tc.mu.txn.ID, newTxn))
```
___
*[pkg/kv/db_test.go, line 841 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvOlhdO9bHvxcCqgr1p:-MvX-Q535nstgELXT-3e:b-sn6ba6) ([raw file](https://github.com/cockroachdb/cockroach/blob/75223e2da057ad8616b9a1ef3992f5da929310f9/pkg/kv/db_test.go#L841)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we assert on the contents of the error after we've checked that it's non-nil?
</blockquote></details>

I added
require.Equal(t, txn.ID(), pErr.TxnID)
I'm happy to add more..
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 323 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvX9fgd51Wop4wDL31O:-MvX9fgd51Wop4wDL31P:byzlt6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d542321fa728b6ed9fb0001075da0d5d5dcde8/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L323)):*
> ```Go
> 		// happens due to a heartbeat failure. On the other hand
> 		// GetLeafTxnFinalState() calls maybeRejectClientLocked() without trying a
> 		// Send().
> ```

Please take a look at this comment Nathan, I forgot this when we talked - this is why Get() doesn't work here. We can do a sleep and then Get() but this is cleaner.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
This <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, although I think we can clean up the test just a bit further!

> Reviewed 13 of 13 files at r7, 12 of 13 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-:-MvesHZV5mpR4u3ntY-a:b-ozvvbi)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @lidorcarmel, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 272 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-Mveqimb1ZBFh2cmFrWG:-Mveqimb1ZBFh2cmFrWH:b531mrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d542321fa728b6ed9fb0001075da0d5d5dcde8/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L272)):*
> ```Go
> 	// Create a DB with a short heartbeat interval.
> 	s := createTestDBWithKnobs(t, &kvserver.StoreTestingKnobs{
> 		DisableScanner:    true,
> ```

Do we need these three knobs?
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 323 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvX9fgd51Wop4wDL31O:-MveqzZF8q0Q0zhB4pBh:bz5ah0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d542321fa728b6ed9fb0001075da0d5d5dcde8/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L323)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

Please take a look at this comment Nathan, I forgot this when we talked - this is why Get() doesn't work here. We can do a sleep and then Get() but this is cleaner.
</blockquote></details>

Interesting. I wasn't expecting the `Get` to notice that the txn was aborted and return an error. This must be because we check the abort span during the read, which is populated on the target range when the high-priority txn aborts our intent. See [WriteAbortSpanOnResolve ](https://github.com/cockroachdb/cockroach/blob/657c0e14884a7c29cf94192da14ccba699a555aa/pkg/kv/kvserver/batcheval/transaction.go#L58) and the `pkg/kv/kvserver/abortspan` package for a bit of an explanation. The high-level role of the "abort span" is to prevent zombie transactions from failing to read their own writes after those writes have been aborted.

One way to get around this is to send the put and the get to different ranges. For instance, you could perform a split on key "b" (`keyB := roachpb.Key("b")`) using `s.DB.AdminSplit` and then perform `Get("b")` requests in this `SucceedsSoon` loop. What do you think about trying that out?
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 336 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-Mves2yS57_Cd3QEi-Qt:-Mves2yS57_Cd3QEi-Qu:buxyfus) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d542321fa728b6ed9fb0001075da0d5d5dcde8/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L336)):*
> ```Go
> 	pErr := tc.GetTxnRetryableErr(ctx)
> 	require.NotNil(t, pErr)
> 	require.Equal(t, txn.ID(), pErr.TxnID)
> ```

Should we assert the error type here as well?


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
yep, thanks, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-:-MvfUu4LCNvrTwrAUNmJ:b-puub2p)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 272 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-Mveqimb1ZBFh2cmFrWG:-MvfNASRB0A_e154u3p3:b-kj8xw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d542321fa728b6ed9fb0001075da0d5d5dcde8/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L272)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need these three knobs?
</blockquote></details>

No, removed.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 323 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-MvX9fgd51Wop4wDL31O:-MvfUTtVDI5KUAaT1W11:b-r9a4de) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d542321fa728b6ed9fb0001075da0d5d5dcde8/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L323)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Interesting. I wasn't expecting the `Get` to notice that the txn was aborted and return an error. This must be because we check the abort span during the read, which is populated on the target range when the high-priority txn aborts our intent. See [WriteAbortSpanOnResolve ](https://github.com/cockroachdb/cockroach/blob/657c0e14884a7c29cf94192da14ccba699a555aa/pkg/kv/kvserver/batcheval/transaction.go#L58) and the `pkg/kv/kvserver/abortspan` package for a bit of an explanation. The high-level role of the "abort span" is to prevent zombie transactions from failing to read their own writes after those writes have been aborted.

One way to get around this is to send the put and the get to different ranges. For instance, you could perform a split on key "b" (`keyB := roachpb.Key("b")`) using `s.DB.AdminSplit` and then perform `Get("b")` requests in this `SucceedsSoon` loop. What do you think about trying that out?
</blockquote></details>

Done, works like a charm.
I did like my "creative" approach but this is probably better, thanks for the details!
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 336 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-Mves2yS57_Cd3QEi-Qt:-MvfODn0DG3NdISYGCFN:bmhwwrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d542321fa728b6ed9fb0001075da0d5d5dcde8/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L336)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we assert the error type here as well?
</blockquote></details>

so, GetTxnRetryableErr() returns *roachpb.TransactionRetryWithProtoRefreshError, not "error", so the type will always be fine, right? or maybe I'm missing something..


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> feel free to merge if/when you feel good about the last two comments. No need for another round of reviews.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-:-Mvu0tNwDOSI6RJGmcEo:b-lwujix)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @lidorcarmel, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 336 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-Mves2yS57_Cd3QEi-Qt:-Mvu0XPq0y6_CP33xqE4:bo94act) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d542321fa728b6ed9fb0001075da0d5d5dcde8/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L336)):*
<details><summary><i>Previously, lidorcarmel (Lidor Carmel) wrote…</i></summary><blockquote>

so, GetTxnRetryableErr() returns *roachpb.TransactionRetryWithProtoRefreshError, not "error", so the type will always be fine, right? or maybe I'm missing something..
</blockquote></details>

Ah, ok, then maybe just add an assertion that `pErr.Transaction.ID` is not equal to `pErr.TxnID`, which indicates an aborted txn.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 300 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-Mvu0LNnAyKRMLMhdxe2:-Mvu0LNnAyKRMLMhdxe3:b-x8ho81) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b5eea715492ef0aeee41e238a672250325145be/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L300)):*
> ```Go
> 	// Note that if we read from the same range then we will cause the txn to be
> 	// poisoned immediately (Send() will fail), even before the heartbeat failure,
> 	// which is not the case we want to test here.
> ```

Let's mention the abort span here to explain why we need to send the Get to a different range.


<!-- Sent from Reviewable.io -->

lidorcarmel reviewed (commented):
Thanks! done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-:-MvurGWuBeSsxsT3XpdP:b4pqkov)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 336 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-Mves2yS57_Cd3QEi-Qt:-Mvuj9g51GULuY40SbQl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d542321fa728b6ed9fb0001075da0d5d5dcde8/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L336)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Ah, ok, then maybe just add an assertion that `pErr.Transaction.ID` is not equal to `pErr.TxnID`, which indicates an aborted txn.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go, line 300 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74563#-Mvu0LNnAyKRMLMhdxe2:-MvumbX95DeXYKqJcOzx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b5eea715492ef0aeee41e238a672250325145be/pkg/kv/kvclient/kvcoord/txn_coord_sender_test.go#L300)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's mention the abort span here to explain why we need to send the Get to a different range.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

lidorcarmel commented:
TFTR Nathan!
bors r+
